pilot/pkg/bootstrap/certcontroller.go

```diff
@@ -205,6 +222,28 @@ func (s *Server) watchRootCertAndGenKeyCert(stop <-chan struct{}) {
 	}
 }
 
+func (s *Server) RotateDNSCertForK8sCA(stop <-chan struct{},
+	defaultCACertPath string,
+	signerName string,
+	approveCsr bool,
+	requestedLifetime time.Duration,
+) {
+	certUtil := certutil.NewCertUtil(int(defaultCertGracePeriodRatio * 100))
```

Comment:
does this mean we only refresh when its already expired? Shouldn't we refresh before its expired?

Reply:
It will be refreshed according to grace period ratio before expiry, for this case grace period ratio is 0.5 so will be refreshed in half time of TTL.